File: #1. 모의고사 2회/장효빈/problem1_실습용_로봇.py
# 풀이 시간: 10분? 기억이 안 남 
# 시간 복잡도: O(N)
# 코드 설명: 명령에 맞게 방향을 전환하거나 이동함. d를 간단히 조작해서 코드를 간단화 할 수 있음.
# 취약한 부분: 해당 문제는 x, y의 좌표평면임. 따라서 r, c로 하기보다는 x, y에 맞게 좌표와 dx, dy를 설정하는 것이 좋아보임.

def solution(command):
    dcs, drs = (0, 1, 0, -1), (1, 0, -1, 0)
    
    r, c, d = 0, 0, 0
    for com in command:
        if com == 'R':
            d = (d + 1) % 4
        elif com == 'L':
            d = (d + 3) % 4
        elif com == 'G':
            r += drs[d]
            c += dcs[d]
        elif com == 'B':
            r, c = r - drs[d], c - dcs[d]
        
    answer = [c, r]
    return answer

File: #1. 모의고사 2회/진성범/problem1_실습용로봇.java
//풀이 시간 : 30분
//시간복잡도 : O(N)
//코드 설명 : 모든 케이스 다 나눠서 일일히 노가다
//취약한 부분 : 더 간략하게 풀어야 할 것 같음

package PCCP;

class problem1_실습용로봇 {
    public int[] solution(String command) {
        int[] answer = {0, 0};
        char dir = 'y';
        boolean plus = true;
        for(int i = 0; i < command.length(); i++) {
            char lett = command.charAt(i);
            if(lett == 'G') {
                if(dir == 'x') {
                    if(plus) {
                        answer[0]++;
                    } else {
                        answer[0]--;
                    }
                } else if(dir == 'y') {
                    if(plus) {
                        answer[1]++;
                    } else {
                        answer[1]--;
                    }
                }
            } else if(lett == 'R') {
                if(dir == 'x') {
                    if(plus) {
                        dir = 'y';
                        plus = false;
                    } else {
                        dir = 'y';
                        plus = true;
                    }
                } else if(dir == 'y') {
                    dir = 'x';
                }
            } else if(lett == 'L') {
                if(dir == 'x') {
                    dir = 'y';
                } else if(dir == 'y') {
                    if(plus) {
                        dir = 'x';
                        plus = false;
                    } else {
                        dir = 'x';
                        plus = true;
                    }
                }
            } else if(lett == 'B') {
                if(dir == 'x') {
                    if(plus) {
                        answer[0]--;
                    } else {
                        answer[0]++;
                    }
                } else if(dir == 'y') {
                    if(plus) {
                        answer[1]--;
                    } else {
                        answer[1]++;
                    }
                }
            }
        }
        return answer;
    }
}
